<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Stacked C+T (SCT) — SCT • bigsnpr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Stacked C+T (SCT) — SCT" />
<meta property="og:description" content="Polygenic Risk Scores for a grid of clumping and thresholding parameters.
Stacking over many Polygenic Risk Scores, corresponding to a grid of many
different parameters for clumping and thresholding." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Stacked C+T (SCT)</h1>
    <small class="dont-index">Source: <a href='https://github.com/privefl/bigsnpr/blob/master/R/SCT.R'><code>R/SCT.R</code></a></small>
    <div class="hidden name"><code>SCT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Polygenic Risk Scores for a grid of clumping and thresholding parameters.</p>
<p>Stacking over many Polygenic Risk Scores, corresponding to a grid of many
different parameters for clumping and thresholding.</p>
    </div>

    <pre class="usage"><span class='fu'>snp_grid_clumping</span>(
  <span class='no'>G</span>,
  <span class='no'>infos.chr</span>,
  <span class='no'>infos.pos</span>,
  <span class='no'>lpS</span>,
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>),
  <span class='kw'>grid.thr.r2</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>0.01</span>, <span class='fl'>0.05</span>, <span class='fl'>0.1</span>, <span class='fl'>0.2</span>, <span class='fl'>0.5</span>, <span class='fl'>0.8</span>, <span class='fl'>0.95</span>),
  <span class='kw'>grid.base.size</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>50</span>, <span class='fl'>100</span>, <span class='fl'>200</span>, <span class='fl'>500</span>),
  <span class='kw'>infos.imp</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span>(<span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span>(<span class='no'>G</span>)),
  <span class='kw'>grid.thr.imp</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>groups</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='fu'>cols_along</span>(<span class='no'>G</span>)),
  <span class='kw'>exclude</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>snp_grid_PRS</span>(
  <span class='no'>G</span>,
  <span class='no'>all_keep</span>,
  <span class='no'>betas</span>,
  <span class='no'>lpS</span>,
  <span class='kw'>n_thr_lpS</span> <span class='kw'>=</span> <span class='fl'>50</span>,
  <span class='kw'>grid.lpS.thr</span> <span class='kw'>=</span> <span class='fl'>0.9999</span> * <span class='fu'><a href='seq_log.html'>seq_log</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span>(<span class='fl'>0.1</span>, <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>min</a></span>(<span class='no'>lpS</span>, <span class='kw'>na.rm</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)), <span class='fu'><a href='https://rdrr.io/r/base/Extremes.html'>max</a></span>(<span class='no'>lpS</span>, <span class='kw'>na.rm</span> <span class='kw'>=</span>
    <span class='fl'>TRUE</span>), <span class='no'>n_thr_lpS</span>),
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>),
  <span class='kw'>backingfile</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span>(),
  <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"float"</span>, <span class='st'>"double"</span>),
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>snp_grid_stacking</span>(
  <span class='no'>multi_PRS</span>,
  <span class='no'>y.train</span>,
  <span class='kw'>alphas</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>1</span>, <span class='fl'>0.01</span>, <span class='fl'>1e-04</span>),
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>G</th>
      <td><p>A FBM.code256
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br />
<strong>You shouldn't have missing values.</strong> Also, remember to do quality control,
e.g. some algorithms in this package won't work if you use SNPs with 0 MAF.</p></td>
    </tr>
    <tr>
      <th>infos.chr</th>
      <td><p>Vector of integers specifying each SNP's chromosome.<br />
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></td>
    </tr>
    <tr>
      <th>infos.pos</th>
      <td><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br />
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></td>
    </tr>
    <tr>
      <th>lpS</th>
      <td><p>Numeric vector of <code>-log10(p-value)</code> associated with <code>betas</code>.</p></td>
    </tr>
    <tr>
      <th>ind.row</th>
      <td><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>grid.thr.r2</th>
      <td><p>Grid of thresholds over the squared correlation between
two SNPs for clumping. Default is <code><a href='https://rdrr.io/r/base/c.html'>c(0.01, 0.05, 0.1, 0.2, 0.5, 0.8, 0.95)</a></code>.</p></td>
    </tr>
    <tr>
      <th>grid.base.size</th>
      <td><p>Grid for base window sizes. Sizes are then computed as
<code>base.size / thr.r2</code> (in kb). Default is <code><a href='https://rdrr.io/r/base/c.html'>c(50, 100, 200, 500)</a></code>.</p></td>
    </tr>
    <tr>
      <th>infos.imp</th>
      <td><p>Vector of imputation scores. Default is all <code>1</code> if you do
not provide it.</p></td>
    </tr>
    <tr>
      <th>grid.thr.imp</th>
      <td><p>Grid of thresholds over <code>infos.imp</code> (default is <code>1</code>), but
you should change it (e.g. <code><a href='https://rdrr.io/r/base/c.html'>c(0.3, 0.6, 0.9, 0.95)</a></code>) if providing <code>infos.imp</code>.</p></td>
    </tr>
    <tr>
      <th>groups</th>
      <td><p>List of vectors of indices to define your own categories.
This could be used e.g. to derive C+T scores using two different GWAS
summary statistics, or to include other information such as functional
annotations. Default just makes one group with all variants.</p></td>
    </tr>
    <tr>
      <th>exclude</th>
      <td><p>Vector of SNP indices to exclude anyway.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    <tr>
      <th>all_keep</th>
      <td><p>Output of <code>snp_grid_clumping()</code> (indices passing clumping).</p></td>
    </tr>
    <tr>
      <th>betas</th>
      <td><p>Numeric vector of weights (effect sizes from GWAS) associated
with each variant (column of <code>G</code>). If alleles are reversed, make sure to
multiply corresponding effects by <code>-1</code>.</p></td>
    </tr>
    <tr>
      <th>n_thr_lpS</th>
      <td><p>Length for default <code>grid.lpS.thr</code>. Default is <code>50</code>.</p></td>
    </tr>
    <tr>
      <th>grid.lpS.thr</th>
      <td><p>Sequence of thresholds to apply on <code>lpS</code>.
Default is a grid (of length <code>n_thr_lpS</code>) evenly spaced on a logarithmic
scale, i.e. on a log-log scale for p-values.</p></td>
    </tr>
    <tr>
      <th>backingfile</th>
      <td><p>Prefix for backingfiles where to store scores of C+T.
As we typically use a large grid, this can result in a large matrix so that
we store it on disk. Default uses a temporary file.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>Type of backingfile values. Either <code>"float"</code> (the default) or
<code>"double"</code>. Using <code>"float"</code> requires half disk space.</p></td>
    </tr>
    <tr>
      <th>multi_PRS</th>
      <td><p>Output of <code>snp_grid_PRS()</code>.</p></td>
    </tr>
    <tr>
      <th>y.train</th>
      <td><p>Vector of phenotypes. If there are two levels (binary 0/1),
it uses <code>big_spLogReg()</code> for stacking, otherwise <code>big_spLinReg()</code>.</p></td>
    </tr>
    <tr>
      <th>alphas</th>
      <td><p>Vector of values for grid-search. See <code>big_spLogReg()</code>.
Default for this function is <code><a href='https://rdrr.io/r/base/c.html'>c(1, 0.01, 0.0001)</a></code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Other parameters to be passed to <code>big_spLogReg()</code>. For example,
using <code>covar.train</code>, you can add covariates in the model with all C+T scores.
You can also use <code>pf.covar</code> if you do not want to penalize these covariates.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>snp_grid_PRS()</code>: An <code>FBM</code> (matrix on disk) that stores the C+T scores
for all parameters of the grid (and for each chromosome separately).
It also stores as attributes the input parameters <code>all_keep</code>, <code>betas</code>,
<code>lpS</code> and <code>grid.lpS.thr</code> that are also needed in <code>snp_grid_stacking()</code>.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


