<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>LD clumping — bed_clumping • bigsnpr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="LD clumping — bed_clumping" />
<meta property="og:description" content="For a bigSNP:
snp_pruning(): LD pruning. Similar to &quot;--indep-pairwise (size+1) 1 thr.r2&quot;
in PLINK.
This function is deprecated (see this article).
snp_clumping() (and bed_clumping()): LD clumping. If you do not provide
any statistic to rank SNPs, it would use minor allele frequencies (MAFs),
making clumping similar to pruning.
snp_indLRLDR(): Get SNP indices of long-range LD regions for the
human genome.

" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>LD clumping</h1>
    <small class="dont-index">Source: <a href='https://github.com/privefl/bigsnpr/blob/master/R/bed-clumping.R'><code>R/bed-clumping.R</code></a>, <a href='https://github.com/privefl/bigsnpr/blob/master/R/clumping.R'><code>R/clumping.R</code></a></small>
    <div class="hidden name"><code>snp_clumping.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a <code>bigSNP</code>:</p><ul>
<li><p><code>snp_pruning()</code>: LD pruning. Similar to "<code>--indep-pairwise (size+1) 1 thr.r2</code>"
in <a href='https://www.cog-genomics.org/plink/1.9/ld'>PLINK</a>.
<strong>This function is deprecated (see <a href='http://bit.ly/2uKo3MN'>this article</a>).</strong></p></li>
<li><p><code>snp_clumping()</code> (and <code>bed_clumping()</code>): LD clumping. If you do not provide
any statistic to rank SNPs, it would use minor allele frequencies (MAFs),
making clumping similar to pruning.</p></li>
<li><p><code>snp_indLRLDR()</code>: Get SNP indices of long-range LD regions for the
human genome.</p></li>
</ul>

    </div>

    <pre class="usage"><span class='fu'>bed_clumping</span>(
  <span class='no'>obj.bed</span>,
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>obj.bed</span>),
  <span class='kw'>S</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>thr.r2</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>100</span>/<span class='no'>thr.r2</span>,
  <span class='kw'>exclude</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>snp_clumping</span>(
  <span class='no'>G</span>,
  <span class='no'>infos.chr</span>,
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>),
  <span class='kw'>S</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>thr.r2</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>100</span>/<span class='no'>thr.r2</span>,
  <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>is.size.in.bp</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>exclude</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>snp_pruning</span>(
  <span class='no'>G</span>,
  <span class='no'>infos.chr</span>,
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>G</span>),
  <span class='kw'>size</span> <span class='kw'>=</span> <span class='fl'>49</span>,
  <span class='kw'>is.size.in.bp</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>thr.r2</span> <span class='kw'>=</span> <span class='fl'>0.2</span>,
  <span class='kw'>exclude</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>nploidy</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)

<span class='fu'>snp_indLRLDR</span>(<span class='no'>infos.chr</span>, <span class='no'>infos.pos</span>, <span class='kw'>LD.regions</span> <span class='kw'>=</span> <span class='no'>LD.wiki34</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>obj.bed</th>
      <td><p>Object of type <code>bed</code>, which is the mapping of some bed file.
Use <code>obj.bed &lt;- bed(bedfile)</code> to get this object.</p></td>
    </tr>
    <tr>
      <th>ind.row</th>
      <td><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>S</th>
      <td><p>A vector of column statistics which express the importance
of each SNP (the more important is the SNP, the greater should be
the corresponding statistic).<br />
For example, if <code>S</code> follows the standard normal distribution, and "important"
means significantly different from 0, you must use <code><a href='https://rdrr.io/r/base/MathFun.html'>abs(S)</a></code> instead.<br />
<strong>If not specified, MAFs are computed and used.</strong></p></td>
    </tr>
    <tr>
      <th>thr.r2</th>
      <td><p>Threshold over the squared correlation between two SNPs.
Default is <code>0.2</code>.</p></td>
    </tr>
    <tr>
      <th>size</th>
      <td><p>For one SNP, window size around this SNP to compute correlations.
Default is <code>100 / thr.r2</code> for clumping (0.2 -&gt; 500; 0.1 -&gt; 1000; 0.5 -&gt; 200).
If not providing <code>infos.pos</code> (<code>NULL</code>, the default), this is a window in
number of SNPs, otherwise it is a window in kb (genetic distance).
I recommend that you provide the positions if available.</p></td>
    </tr>
    <tr>
      <th>exclude</th>
      <td><p>Vector of SNP indices to exclude anyway. For example,
can be used to exclude long-range LD regions (see Price2008). Another use
can be for thresholding with respect to p-values associated with <code>S</code>.</p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    <tr>
      <th>G</th>
      <td><p>A FBM.code256
(typically <code>&lt;bigSNP&gt;$genotypes</code>).<br />
<strong>You shouldn't have missing values.</strong> Also, remember to do quality control,
e.g. some algorithms in this package won't work if you use SNPs with 0 MAF.</p></td>
    </tr>
    <tr>
      <th>infos.chr</th>
      <td><p>Vector of integers specifying each SNP's chromosome.<br />
Typically <code>&lt;bigSNP&gt;$map$chromosome</code>.</p></td>
    </tr>
    <tr>
      <th>infos.pos</th>
      <td><p>Vector of integers specifying the physical position
on a chromosome (in base pairs) of each SNP.<br />
Typically <code>&lt;bigSNP&gt;$map$physical.pos</code>.</p></td>
    </tr>
    <tr>
      <th>is.size.in.bp</th>
      <td><p>Deprecated.</p></td>
    </tr>
    <tr>
      <th>nploidy</th>
      <td><p>Number of trials, parameter of the binomial distribution.
Default is <code>2</code>, which corresponds to diploidy, such as for the human genome.</p></td>
    </tr>
    <tr>
      <th>LD.regions</th>
      <td><p>A <code>data.frame</code> with columns "Chr", "Start" and "Stop".
Default use the table of 34 long-range LD regions that you can find
<a href='https://goo.gl/0Ou7uI'>there</a>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p><code>snp_clumping()</code> (and <code>bed_clumping()</code>): SNP indices that are <strong>kept</strong>.</p></li>
<li><p><code>snp_indLRLDR()</code>: SNP indices to be used as (part of) the '<strong><code>exclude</code></strong>'
parameter of <code>snp_clumping()</code>.</p></li>
</ul>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Price AL, Weale ME, Patterson N, et al.
Long-Range LD Can Confound Genome Scans in Admixed Populations.
Am J Hum Genet. 2008;83(1):132-135.
<a href='http://dx.doi.org/10.1016/j.ajhg.2008.06.005'>http://dx.doi.org/10.1016/j.ajhg.2008.06.005</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attachExtdata.html'>snp_attachExtdata</a></span>()
<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>test</span>$<span class='no'>genotypes</span>

<span class='co'># clumping (prioritizing higher MAF)</span>
<span class='no'>ind.keep</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_clumping</span>(<span class='no'>G</span>, <span class='kw'>infos.chr</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>chromosome</span>,
                         <span class='kw'>infos.pos</span> <span class='kw'>=</span> <span class='no'>test</span>$<span class='no'>map</span>$<span class='no'>physical.pos</span>,
                         <span class='kw'>thr.r2</span> <span class='kw'>=</span> <span class='fl'>0.1</span>)

<span class='co'># keep most of them -&gt; not much LD in this simulated dataset</span>
<span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span>(<span class='no'>ind.keep</span>) / <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span>(<span class='no'>G</span>)</div><div class='output co'>#&gt; [1] 0.7919419</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


