<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read PLINK files into a "bigSNP" — snp_readBed • bigsnpr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Read PLINK files into a "bigSNP" — snp_readBed" />
<meta property="og:description" content="Functions to read bed/bim/fam files into a bigSNP." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bigsnpr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.4.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Manual</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/demo.html">Quick demo</a>
    </li>
    <li>
      <a href="../articles/exo.html">Exercise</a>
    </li>
    <li>
      <a href="../articles/pruning-vs-clumping.html">Why clumping should be preferred over pruning</a>
    </li>
    <li>
      <a href="../articles/how-to-PCA.html">How to capture Population Structure with PCA (LD problem explained)</a>
    </li>
    <li>
      <a href="../articles/bedpca.html">How to capture Population Structure with PCA (directly on PLINK bed files)</a>
    </li>
    <li>
      <a href="../articles/SCT.html">Computing polygenic scores using Stacked Clumping and Thresholding (SCT)</a>
    </li>
    <li>
      <a href="../articles/LDpred2.html">Computing polygenic scores using LDpred2</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://privefl.github.io/about.html">
    <span class="fa fa-user"></span>
     
    About
  </a>
</li>
<li>
  <a href="https://github.com/privefl/bigsnpr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read PLINK files into a "bigSNP"</h1>
    <small class="dont-index">Source: <a href='https://github.com/privefl/bigsnpr/blob/master/R/read-plink.R'><code>R/read-plink.R</code></a></small>
    <div class="hidden name"><code>snp_readBed.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Functions to read bed/bim/fam files into a <a href='bigSNP-class.html'>bigSNP</a>.</p>
    </div>

    <pre class="usage"><span class='fu'>snp_readBed</span>(<span class='no'>bedfile</span>, <span class='kw'>backingfile</span> <span class='kw'>=</span> <span class='fu'><a href='sub_bed.html'>sub_bed</a></span>(<span class='no'>bedfile</span>))

<span class='fu'>snp_readBed2</span>(
  <span class='no'>bedfile</span>,
  <span class='kw'>backingfile</span> <span class='kw'>=</span> <span class='fu'><a href='sub_bed.html'>sub_bed</a></span>(<span class='no'>bedfile</span>),
  <span class='kw'>ind.row</span> <span class='kw'>=</span> <span class='fu'>rows_along</span>(<span class='no'>obj.bed</span>),
  <span class='kw'>ind.col</span> <span class='kw'>=</span> <span class='fu'>cols_along</span>(<span class='no'>obj.bed</span>),
  <span class='kw'>ncores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bedfile</th>
      <td><p>Path to file with extension ".bed" to read.
You need the corresponding ".bim" and ".fam" in the same directory.</p></td>
    </tr>
    <tr>
      <th>backingfile</th>
      <td><p>The path (without extension) for the backing files
for the cache of the <a href='bigSNP-class.html'>bigSNP</a> object. Default takes the bedfile
without the ".bed" extension.</p></td>
    </tr>
    <tr>
      <th>ind.row</th>
      <td><p>An optional vector of the row indices (individuals) that
are used. If not specified, all rows are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>ind.col</th>
      <td><p>An optional vector of the column indices (SNPs) that are used.
If not specified, all columns are used.<br />
<strong>Don't use negative indices.</strong></p></td>
    </tr>
    <tr>
      <th>ncores</th>
      <td><p>Number of cores used. Default doesn't use parallelism.
You may use nb_cores.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The path to the RDS file that stores the <code>bigSNP</code> object.
Note that this function creates one other file which stores the values of
the Filebacked Big Matrix.<br />
<strong>You shouldn't read from PLINK files more than once.</strong> Instead, use
<a href='snp_attach.html'>snp_attach</a> to load the "bigSNP" object in any R session from backing files.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>For more information on these formats, please visit
<a href='https://www.cog-genomics.org/plink/1.9/formats#bed'>PLINK webpage</a>.
For other formats, please use PLINK to convert them in bedfiles,
which require minimal space to store and are faster to read. For example,
to convert from a VCF file, use the <code>--vcf</code> option. See <a href='snp_plinkQC.html'>snp_plinkQC</a>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>(<span class='no'>bedfile</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"extdata"</span>, <span class='st'>"example.bed"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bigsnpr"</span>))</div><div class='output co'>#&gt; [1] "C:/Users/au639593/Documents/R/win-library/3.6/bigsnpr/extdata/example.bed"</div><div class='input'>
<span class='co'># Reading the bedfile and storing the data in temporary directory</span>
<span class='no'>rds</span> <span class='kw'>&lt;-</span> <span class='fu'>snp_readBed</span>(<span class='no'>bedfile</span>, <span class='kw'>backingfile</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempfile</a></span>())

<span class='co'># Loading the data from backing files</span>
<span class='no'>test</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='snp_attach.html'>snp_attach</a></span>(<span class='no'>rds</span>)

<span class='fu'><a href='https://rdrr.io/r/utils/str.html'>str</a></span>(<span class='no'>test</span>)</div><div class='output co'>#&gt; List of 3
#&gt;  $ genotypes:Reference class 'FBM.code256' [package "bigstatsr"] with 15 fields
#&gt;   ..$ extptr      :&lt;externalptr&gt; 
#&gt;   ..$ extptr_rw   :&lt;externalptr&gt; 
#&gt;   ..$ nrow        : int 517
#&gt;   ..$ ncol        : int 4542
#&gt;   ..$ type        : Named int 1
#&gt;   .. ..- attr(*, "names")= chr "unsigned char"
#&gt;   ..$ backingfile : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp63sV9P\\file428c5c3a7edf.bk"
#&gt;   ..$ is_read_only: logi FALSE
#&gt;   ..$ address     :&lt;externalptr&gt; 
#&gt;   ..$ address_rw  :&lt;externalptr&gt; 
#&gt;   ..$ bk          : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp63sV9P\\file428c5c3a7edf.bk"
#&gt;   ..$ rds         : chr "C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmp63sV9P\\file428c5c3a7edf.rds"
#&gt;   ..$ is_saved    : logi TRUE
#&gt;   ..$ type_chr    : chr "unsigned char"
#&gt;   ..$ type_size   : int 1
#&gt;   ..$ code256     : num [1:256] 0 1 2 NA NA NA NA NA NA NA ...
#&gt;   ..and 26 methods, of which 12 are  possibly relevant:
#&gt;   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,
#&gt;   ..  check_write_permissions, copy#envRefClass, initialize, initialize#FBM,
#&gt;   ..  save, show#envRefClass, show#FBM
#&gt;  $ fam      :'data.frame':	517 obs. of  6 variables:
#&gt;   ..$ family.ID  : chr [1:517] "POP1" "POP1" "POP1" "POP1" ...
#&gt;   ..$ sample.ID  : chr [1:517] "IND0" "IND1" "IND2" "IND3" ...
#&gt;   ..$ paternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:517] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ affection  : int [1:517] 1 1 2 1 1 1 1 1 1 1 ...
#&gt;  $ map      :'data.frame':	4542 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:4542] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:4542] "SNP0" "SNP1" "SNP2" "SNP3" ...
#&gt;   ..$ genetic.dist: int [1:4542] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:4542] 112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;   ..$ allele1     : chr [1:4542] "A" "T" "T" "T" ...
#&gt;   ..$ allele2     : chr [1:4542] "T" "A" "A" "A" ...
#&gt;  - attr(*, "class")= chr "bigSNP"</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span>(<span class='no'>G</span> <span class='kw'>&lt;-</span> <span class='no'>test</span>$<span class='no'>genotypes</span>)</div><div class='output co'>#&gt; [1]  517 4542</div><div class='input'><span class='no'>G</span>[<span class='fl'>1</span>:<span class='fl'>8</span>, <span class='fl'>1</span>:<span class='fl'>8</span>]</div><div class='output co'>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; [1,]    0    0    2    0    1    1    0    2
#&gt; [2,]    1    0    1    0    0    1    0    2
#&gt; [3,]    0    1    1    0    2    1    0    2
#&gt; [4,]    0    0    2    0    2    1    0    2
#&gt; [5,]    1    0    0    0    2    2    1    0
#&gt; [6,]    0    1    0    0    2    1    0    0
#&gt; [7,]    0    1    1    0    2    2    1    1
#&gt; [8,]    1    0    1    1    1    1    0    1</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Florian Privé.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


